source/common/unified_memory_allocation/src/memory_tracker.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ */
+
+#include "memory_tracker_internal.h"
+
+#include <map>
+#include <mutex>
+#include <shared_mutex>
+
+// TODO: reimplement in C and optimize...
+struct uma_memory_tracker_t {
+    enum uma_result_t add(void *pool, const void *ptr, size_t size) {
+        std::unique_lock<std::shared_mutex> lock(mtx);
+
+        auto ret = map.try_emplace(reinterpret_cast<uintptr_t>(ptr), size, pool);
+        return ret.second ? UMA_RESULT_SUCCESS : UMA_RESULT_ERROR_UNKNOWN;
+    }
+
+    enum uma_result_t remove(const void *ptr, size_t size) {
+        std::unique_lock<std::shared_mutex> lock(mtx);
+
+        map.erase(reinterpret_cast<uintptr_t>(ptr));
+
+        // TODO: handle removing part of the range
+        (void)size;
+
+        return UMA_RESULT_SUCCESS;
+    }
+
+    void *find(const void *ptr) {
+        std::shared_lock<std::shared_mutex> lock(mtx);
+
+        auto intptr = reinterpret_cast<uintptr_t>(ptr);
+        auto it = map.upper_bound(intptr);
+        if (it == map.begin()) {
+            return nullptr;
+        }
+
+        --it;
+        if (intptr >= it->first && intptr < it->first + it->second.first) {
+            return it->second.second;
+        }
+
+        return nullptr;
+    }
+
+  private:
+    std::shared_mutex mtx;
+    std::map<uintptr_t, std::pair<size_t, void *>> map;
+};
+
+extern "C" {
+
+uma_memory_tracker_handle_t umaMemoryTrackerGet() {
+    static uma_memory_tracker_t tracker;
```

Comment:
I'm not sure if I like this being a global.

Reply:
Another option is to expose trackerCreate method to the user, but then the handle needs to be passed trackingMemoryProviderCreate or to poolCreate (if we decide to wrap providers ourselves) and it will mean that different applications using UMA will not share the tracker - this is problematic for cases like MPI.